src/goSuggest.ts

```diff
@@ -66,80 +74,129 @@ export class GoCompletionItemProvider implements vscode.CompletionItemProvider {
+			let gocode = getBinPath('gocode');
+
+			// Unset GOOS and GOARCH for the `gocode` process to ensure that GOHOSTOS and GOHOSTARCH
+			// are used as the target operating system and architecture. `gocode` is unable to provide
+			// autocompletion when the Go environment is configured for cross compilation.
+			let env = Object.assign({}, process.env, { GOOS: '', GOARCH: '' });
+
+			// Spawn `gocode` process
+			let p = cp.execFile(gocode, ['-f=json', 'autocomplete', filename, 'c' + offset], { env }, (err, stdout, stderr) => {
+				try {
+					if (err && (<any>err).code === 'ENOENT') {
+						promptForMissingTool('gocode');
+					}
+					if (err) return reject(err);
+					let results = <[number, GoCodeSuggestion[]]>JSON.parse(stdout.toString());
+					let suggestions = [];
+					// 'Smart Snippet' for package clause
+					// TODO: Factor this out into a general mechanism
+					if (!inputText.match(/package\s+(\w+)/)) {
+						let defaultPackageName =
+							basename(filename) === 'main.go'
+								? 'main'
+								: basename(dirname(filename));
+						let packageItem = new vscode.CompletionItem('package ' + defaultPackageName);
+						packageItem.kind = vscode.CompletionItemKind.Snippet;
+						packageItem.insertText = 'package ' + defaultPackageName + '\r\n\r\n';
+						suggestions.push(packageItem);
+
+					}
+					if (results[1]) {
+						for (let suggest of results[1]) {
+							if (inString && suggest.class !== 'import') continue;
+							let item = new vscode.CompletionItem(suggest.name);
+							item.kind = vscodeKindFromGoCodeClass(suggest.class);
+							item.detail = suggest.type;
+							if (inString && suggest.class === 'import') {
+								item.textEdit = new vscode.TextEdit(
+									new vscode.Range(
+										position.line,
+										lineText.substring(0, position.character).lastIndexOf('"') + 1,
+										position.line,
+										position.character),
+									suggest.name
+								);
+							}
+							let conf = vscode.workspace.getConfiguration('go');
+							if (conf.get('useCodeSnippetsOnFunctionSuggest') && suggest.class === 'func') {
+								let params = parameters(suggest.type.substring(4));
+								let paramSnippets = [];
+								for (let i in params) {
+									let param = params[i].trim();
+									if (param) {
+										param = param.replace('{', '\\{').replace('}', '\\}');
+										paramSnippets.push('{{' + param + '}}');
+									}
+								}
+								item.insertText = suggest.name + '(' + paramSnippets.join(', ') + ') {{}}';
+							}
+							suggestions.push(item);
+						};
+					}
+					resolve(suggestions);
+				} catch (e) {
+					reject(e);
+				}
+			});
+			p.stdin.end(inputText);
+		});
+	}
 	// TODO: Shouldn't lib-path also be set?
 	private ensureGoCodeConfigured(): Thenable<void> {
-		return new Promise<void>((resolve, reject) => {
+		let pkgPromise = listPackages().then((pkgs: string[]) => {
+this.pkgsList = pkgs.map(pkg => {
```

Comment:
Is it okay for this to be fixed at startup time, not up to date as new packages are installed into the GOPATH?